receiver/stats.go

```diff
@@ -0,0 +1,155 @@
+/*
+ * skogul, stats receiver
+ *
+ * Copyright (c) 2019-2020 Telenor Norge AS
+ * Author(s):
+ *  - Håkon Solbjørg <hakon.solbjorg@telenor.no>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301  USA
+ */
+
+package receiver
+
+import (
+	"context"
+	"time"
+
+	"github.com/telenornms/skogul"
+)
+
+var statsLog = skogul.Logger("receiver", "stats")
+
+// Stats receives metrics from skogul and forwards it to a handler.
+type Stats struct {
+	Handler  *skogul.HandlerRef
+	Interval skogul.Duration
+	ChanSize uint64
+	ch       chan *skogul.Metric
+	ticker   *time.Ticker
+}
+
+// statsDrainCtx and statsDrainCancel are the context and cancel functions
+// for the automatically created skogul.StatsChan.
+// If a skogul stats receiver is configured, statsDrainCancel MUST be called
+// so that statistics are not discarded.
+var statsDrainCtx, statsDrainCancel = context.WithCancel(context.Background())
+
+// init makes sure that the skogul stats channel exists at all times.
+// Furthermore, it starts a goroutine to empty the channel in the case
+// that the stats receiver is not configured, in which case the chan
+// would end up blocking after it is filled.
+func init() {
+	// Create skogul.StatsChan so we don't have components blocking on it
+	if skogul.StatsChan == nil {
+		skogul.StatsChan = make(chan *skogul.Metric, 100)
```

Comment:
Experience suggests that any buffer size like this should be configurable, but since we don't provide global configuration (yet), it's a bit of a moot point, I suppose.